### ISSUE TITLE
Binary disassembly and decompilation howto using Ghidra

### ISSUE THREAD

**Issue by nbritton** [osresearch/airbreak#44]
The main chip on my AirSense 10 AutoSet is a STMicroelectronics STM32F405ZGT6. This is an ARM Cortex-M4 32-bit ISA, however the Internet says the Cortex-M series only supports the 16-bit Thumb ISA.

STM32F405ZG Datasheet:
https://www.st.com/resource/en/datasheet/stm32f405zg.pdf

16-bit Thumb ISA Quick Reference:
http://infocenter.arm.com/help/topic/com.arm.doc.qrc0006e/QRC0006_UAL16.pdf

High-res photos of the front and back of the main board in the AirSense 10 AutoSet: http://imgur.com/a/01iV4kP

I was able to successfully disassemble and decompile functions into C source code using Ghidra, below are the steps that I followed:

https://ghidra-sre.org
https://www.oracle.com/java/technologies/javase-jdk14-downloads.html

<img width="1440" alt="Screen Shot 2020-05-03 at 11 36 25 PM" src="https://user-images.githubusercontent.com/22720196/80936635-9f468b00-8d97-11ea-8e35-664fd6edcb70.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 23 23 PM" src="https://user-images.githubusercontent.com/22720196/80936648-aec5d400-8d97-11ea-8553-c6024e14c65b.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 24 09 PM" src="https://user-images.githubusercontent.com/22720196/80936662-bb4a2c80-8d97-11ea-9821-6c750ad2b76d.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 24 36 PM" src="https://user-images.githubusercontent.com/22720196/80936674-c7ce8500-8d97-11ea-967a-1d046bd8a256.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 34 08 PM" src="https://user-images.githubusercontent.com/22720196/80936692-dd43af00-8d97-11ea-8667-971b1c1f93c5.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 34 21 PM" src="https://user-images.githubusercontent.com/22720196/80936705-ea609e00-8d97-11ea-9f07-2f7eb2e065c4.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 24 52 PM" src="https://user-images.githubusercontent.com/22720196/80936720-f77d8d00-8d97-11ea-916c-8d348a1d3f2b.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 25 24 PM" src="https://user-images.githubusercontent.com/22720196/80936735-049a7c00-8d98-11ea-8e90-6207143fda81.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 26 17 PM" src="https://user-images.githubusercontent.com/22720196/80936747-0cf2b700-8d98-11ea-95ad-1dc01c3d01c2.png">

<img width="1440" alt="Screen Shot 2020-05-03 at 11 27 07 PM" src="https://user-images.githubusercontent.com/22720196/80936761-1aa83c80-8d98-11ea-965e-7b1a52ba2ee6.png">

**Comment by bdarcus** [comment]
This is great. But how does one apply this to enhancing the patch script; say adding missing menu items?

https://github.com/osresearch/airbreak/issues/42

**Comment by dblunk88** [comment]
Thank you for the detailed guide, but I also think that you need to set the load address to 0x8000000

**Comment by nbritton** [comment]
Yeah I think you could be right. 08008590 is a data block referenced by a pointer located at 08004128, and PTR_DAT_08004108 is referenced by FUN_08004100:08004100 which returns the pointer data. That appears to make more sense.

Are we able to share the decompiled source code? Just wondering on the best way to collaborate on deconstructing what the functions and parameter data are doing.

At the very least I'd like to figure out how to clear the gibberish that is displayed on my home screen.

**Comment by gszakacs** [comment]
Not sure if I inderstand you correctly, but if if you are referring to gibberish as
![image](https://user-images.githubusercontent.com/10700953/81014279-92bf4280-8e2a-11ea-8313-9772e0cca169.png)
then simply comment out the following two lines in the patch-airsense file:

FLAGSTR=$(printf 'FLAGS=0x%02x' $BUILD_FLAGS)
printf $FLAGSTR | patch 0x17588

COMMIT_HASH=$(git log -n1 --format=format:"%H" | head -c 7)
printf 'GIT=%s\x00' $COMMIT_HASH | patch 0x17764
 
This will leave the button titles with their original values which are:
My options
and
Sleep Report

**Comment by wormyrocks** [comment]
Yes, those options are for debugging purposes - it gives an at-a-glance view of what version of the hacked firmware has been compiled as well as which UI settings have been enabled or disabled. Feel free to comment that out of the patch script.

**Comment by nbritton** [comment]
Ok cool, I didn't realize those were intentional, I just thought it was some gibberish buffer overflow or something. I'm a principal computer systems engineer but this project isn't anywhere near my area of specialization so I don't have a lot of direct experience with reverse engineering.

**Comment by jjsardar** [comment]
how change image in home ?
